handler.go

```diff
@@ -205,229 +189,49 @@ func (fm *Frontman) InputFromHub() (*Input, error) {
-	var errs []string
-	var jsonEncoder = json.NewEncoder(outputFile)
-
-	// encode and add results to the file as we get it from the chan
-	for res := range resultsChan {
-		err := jsonEncoder.Encode(res)
-		if err != nil {
-			errs = append(errs, err.Error())
-		}
-	}
-
-	if errs != nil {
-		return fmt.Errorf("JSON encoding errors: %s", strings.Join(errs, "; "))
-	}
-
-	return nil
-}
-
-func (fm *Frontman) sendResultsChanToFile(resultsChan chan Result, outputFile *os.File) error {
-	var results []Result
-	var jsonEncoder = json.NewEncoder(outputFile)
-	for res := range resultsChan {
-		results = append(results, res)
-	}
-
-	return jsonEncoder.Encode(results)
-}
-
-func (fm *Frontman) sendResultsChanToHub(resultsChan chan Result) error {
-	var results []Result
-	for res := range resultsChan {
-		results = append(results, res)
-	}
-	err := fm.PostResultsToHub(results)
-	if err != nil {
-		return fmt.Errorf("PostResultsToHub: %s", err.Error())
-	}
-
-	fm.Stats.CheckResultsSentToHub += uint64(len(results))
-
-	fm.offlineResultsBuffer = []Result{}
-	return nil
-}
-
-func (fm *Frontman) sendResultsChanToHubWithInterval(resultsChan chan Result) error {
-	sendResultsTicker := time.NewTicker(secToDuration(fm.Config.SenderModeInterval))
-	defer sendResultsTicker.Stop()
-
-	var results []Result
-	shouldReturn := false
+func (fm *Frontman) Run(inputFilePath string, outputFile *os.File, interrupt chan struct{}) {
+	fm.Stats.StartedAt = time.Now()
+	logrus.Debugf("Start writing stats file: %s", fm.Config.StatsFile)
+	fm.StartWritingStats()
 
 	for {
-		select {
-		case res, ok := <-resultsChan:
-			if !ok {
-				// chan was closed
-				// no more results left
-				shouldReturn = true
-				break
+		if err := fm.HealthCheck(); err != nil {
+			fm.HealthCheckPassedPreviously = false
+			logrus.WithError(err).Errorln("Health checks are not passed. Skipping other checks.")
+			select {
+			case <-interrupt:
+				return
+			case <-time.After(secToDuration(fm.Config.Sleep)):
+				continue
 			}
-
-			results = append(results, res)
-			// skip PostResultsToHub
-			continue
-		case <-sendResultsTicker.C:
-			break
-		}
-
-		logrus.Debugf("SenderModeInterval: send %d results", len(results))
-		err := fm.PostResultsToHub(results)
-		if err != nil {
-			err = fmt.Errorf("PostResultsToHub error: %s", err.Error())
-		}
-
-		if shouldReturn {
-			return err
+		} else if !fm.HealthCheckPassedPreviously {
+			fm.HealthCheckPassedPreviously = true
+			logrus.Infoln("All health checks are positive. Resuming normal operation.")
 		}
 
-		if err != nil {
+		input, err := fm.FetchInput(inputFilePath)
+		switch {
+		case err != nil && err == ErrorMissingHubOrInput:
+			logrus.Warnln(err)
+			// This is necessary because MSI does not respect if service quits with status 0 but quickly.
+			// In other cases this delay doesn't matter, but also can be useful for polling config changes in a loop.
+			time.Sleep(10 * time.Second)
+			os.Exit(0)
```

Comment:
Not sure if it's new code or some old code moved to here, but I think calling os.Exit can be dangerous here, as it will not invoke defer calls that might have been declared earlier.

Reply:
Agree, but yes this is old code, just got moved around a bit in this PR

https://github.com/cloudradar-monitoring/frontman/blob/5573900c0d18ac95dfd7372250f33f836a3b5b51/handler.go#L533

Reply:
Also, I don't understand the comment above this code- comment suggests it is nessecary to exit.

Reply:
@nikita-vanyasin could you please review the patchset?